### PR TITLE
Synchronize krnlmon and stratum threads

### PR DESCRIPTION
**Pull request by nupuruttarwar**
Start krnlmon init and shutdown thread after receiving appropriate signal from stratum thread signaling that P4RT servers are started
Signed-off-by: Nupur Uttarwar <nupur.uttarwar@intel.com>

### CODE REVIEW

File: krnlmon_main.h [hunk 1]
```diff
@@ -13,7 +13,7 @@
 extern "C" {
 #endif	
 int krnlmon_init(void);
```
Comment (ffoulkes) [inline krnlmon_main.h:15]: 1. The header file is not dependent on any of the definitions in `pthread.h`, so it should not #include it. This is the responsibility of the [source] files that require pthread definitions. This principle is called "include what you use."
2. I recommend adding a blank lines before and after the function prototypes, separating them from the __cplusplus #ifdef blocks. It makes the file easier to read.

File: switchlink/switchlink_main.c [hunk 2]
```diff
@@ -294,6 +305,13 @@ struct nl_sock *switchlink_get_nl_sock(void) {
 }
 
 void *switchlink_main(void *args) {
+  pthread_mutex_lock(&rpc_start_lock);
+  while(!rpc_start_cookie) {
```
Comment (ffoulkes) [inline switchlink/switchlink_main.c:309]: You might want to add a space between `while` and `(`. It's a control statement, not a function call.

File: switchlink/switchlink_main.c [hunk 3]
```diff
@@ -321,12 +339,16 @@ void *switchlink_main(void *args) {
   return NULL;
 }
 
-int switchlink_stop(void) {
+void *switchlink_stop(void *args) {
+  pthread_mutex_lock(&rpc_stop_lock);
+  while(!rpc_stop_cookie) {
```
Comment (ffoulkes) [inline switchlink/switchlink_main.c:344]: You might want to add a space between `while` and `(`. It's a control statement, not a function call.

File: switchlink/switchlink_main.c [hunk 4]
```diff
@@ -32,6 +32,17 @@ static pthread_mutex_t cookie_mutex;
 static pthread_cond_t cookie_cv;
 static int cookie = 0;
 
+//Switchlink_main pthread variables
+extern pthread_cond_t rpc_start_cond;
+extern pthread_mutex_t rpc_start_lock;
+extern int rpc_start_cookie;
+
+
+//Switchlink stop pthread varaibales
```
Comment (ffoulkes) [inline switchlink/switchlink_main.c:41]: Typo: *variables*

**Comment by ffoulkes** [comment]
I would suggest a short commit title followed by a longer commit comment:

```
Synchronize krnlmon and stratum threads

Start krnlmon init and shutdown thread after receiving appropriate signal from
stratum thread signaling that P4RT servers are started.
```

Also, can you find a clearer wording for the long comment?

File: krnlmon_main.h (hunk 1 above)
Comment (ffoulkes) [reply krnlmon_main.h:15]: Incidentally, the symbol to specify before including pthread.h should be `__USE_GNU`, not `_GNU_SOURCE`.

Comment (nupuruttarwar) [reply krnlmon_main.h:11]: Done

File: krnlmon_main.c [hunk 5]
```diff
@@ -2,26 +2,36 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include <stdio.h>
+#if !defined(_GNU_SOURCE)
+#define __USE_GNU
+#endif
 
+#include <pthread.h>
 #include "krnlmon_main.h"
```
Comment (ffoulkes) [inline krnlmon_main.c:10]: One generally includes one's own header file before any others. It's a quick-and-dirty check for missing dependencies in the header file.
```
#include "krnlmon_main.h"

#include <pthread.h>
#include <stdio.h>
```

**Review by ffoulkes — APPROVED** [review]

File: krnlmon_main.c (hunk 5 above)
Comment (nupuruttarwar) [reply krnlmon_main.c]: Done

**Pull request merged**